### DISCUSSION TITLE
How to produce two PCM audio outputs with different sample rate.

### DISCUSSION THREAD

**Discussion by nikolaypavlov** [python-ffmpegio/python-ffmpegio#51] [Q&A]
I would like to convert mp3 audio into two PCM outputs of different sampling rate, like in the following `ffmpeg` example:

```console
ffmpeg -i audio.mp3 -ar 16000 -acodec pcm_s16le -f s16le -ac 1 audio_16000.pcm -ar 8000 -acodec pcm_s16le -f s16le -ac 1 audio_8000.pcm 
```

What's the best way to implement this?

**Comment by tikuma-lsuhsc** [comment]
@nikolaypavlov - `ffmpegio.transcode()` supports multiple outputs like this:

```python
import ffmpegio as ff

input = 'audio.mp3'

output1 = 'audio_16000.pcm'
output2 = 'audio_8000.pcm'

ff.transcode(input,[(output1,{'ar': 16000},(output2,{'ar': 8000})], acodec='pcm_s16le', f='s16le', ac=1)
```

Basically, feed all the common (or default) options as keyword arguments and specify the output files as a list. The one with the default output options maybe just a file path string and those with specific options, create a tuple list item with file path and its option dict.

P.S., the same can be done with its input argument to feed multiple inputs.

**Reply by nikolaypavlov** [reply]
@tikuma-lsuhsc Thank you. Is this possible to get both outputs as `numpy` arrays instead of files?

**Reply by tikuma-lsuhsc** [reply]
@nikolaypavlov  - Unfortunately no, using the `ffmpegio` top-level API. I'd just call `audio.read()` twice if I were you:

```python
fs1,x1 = ff.audio.read(input, ar=16000, sample_fmt='s16', ac=1)
fs2,x2 = ff.audio.read(input, ar=8000, sample_fmt='s16', ac=1)
```

Be aware that `x1` and `x2` are both "column vectors" (2D array with the second dimension = 1). (I'm planning a next release to allow more flexibility, though)

BTW, if you want normalized signal, you can use `sample_fmt='dbl'` or `'flt'`.

If you're on Linux/Mac, you can make use of `pass_fds` argument of `subprocess.Popen` to achieve this with one subprocess run. But, for audio read operation, it's not worth the effort IMHO.

**Reply by tikuma-lsuhsc** [reply]
Oh, FYI, if you already have your input signal as NumPy array, you can use `ff.audio.filter()` instead, replacing the input file path with sampling rate and numpy array. 

Edit: Actually, you may need to pass a null audio filter `'anull'` as that bug has not been patched yet. If you want to go this route, let me know and I can give you more specific code sample.

**Reply by nikolaypavlov** [reply]
@tikuma-lsuhsc I have them as audio bytes in memory in mp3 format, so not exactly a `numpy` array. I'd like to avoid double decoding as much as possible, because I have a few millions mp3s daily on my service.

**Reply by tikuma-lsuhsc** [reply]
@nikolaypavlov - Then you need to cook up a multi-pipe solution yourself, I'm afraid. `ffmpegio` can take an encoded byte sequence as an input instead of a file, but it currently does not support multiple (numpy) outputs as you desire. 

If you're application is not running on Windows, then as I mentioned you should be able to make use of `pass_fds` argument of `subprocess.Popen` (and all its derivatives) and use the FFmpeg pipe protocol `'pipe:3'` and `pipe:4` (or `pipe:1` and `pipe:3` if you wish to use stdout). You can get data in/out using [`io.BytesIO`](https://docs.python.org/3/library/io.html#io.BytesIO) object.

If it must run on Windows (which doesn't support `pass_fds`) you can play with my named pipe package [`namedpipe`](https://github.com/python-ffmpegio/python-namedpipe). This is a low-level package so you need to build your own thread functions around it.

Good luck. If you need further help, post your code, and I'll be glad to give you further pointers if you're going along one of the above routes.
